swift/ql/test/library-tests/regex/redos_variants.swift

```diff
@@ -0,0 +1,570 @@
+    _ = try Regex("^((x([^Y]+)?)*(Y|$))").firstMatch(in: tainted) // $ SPURIOUS: redos-vulnerable=
+
+    // BAD
+    // (no confirmed attack string)
+    _ = try Regex(#"foo([\w-]*)+bar"#).firstMatch(in: tainted) // $ redos-vulnerable=
+
+    // BAD
+    // attack string: "ab" x lots + "!"
+    _ = try Regex("((ab)*)+c").firstMatch(in: tainted) // $ redos-vulnerable=
+
+    // BAD
+    // attack string: "a" x lots + "!"
+    _ = try Regex("(a?a?)*b").firstMatch(in: tainted) // $ redos-vulnerable=
+
+    // GOOD
+    _ = try Regex("(a?)*b").firstMatch(in: tainted)
+
+    // BAD - but not detected
+    // (no confirmed attack string)
+    _ = try Regex("(c?a?)*b").firstMatch(in: tainted) // $ MISSING: redos-vulnerable=
+
+    // BAD
+    // attack string: "a" x lots + "!"
+    _ = try Regex("(?:a|a?)+b").firstMatch(in: tainted) // $ redos-vulnerable=
+
+    // BAD - but not detected.
+    // attack string: "ab" x lots + "!"
+    _ = try Regex("(a?b?)*$").firstMatch(in: tainted) // $ MISSING: redos-vulnerable=
+
+    // BAD
+    // (no confirmed attack string)
+    _ = try Regex("PRE(([a-c]|[c-d])T(e?e?e?e?|X))+(cTcT|cTXcTX$)").firstMatch(in: tainted) // $ redos-vulnerable=
+
+    // BAD
+    // attack string: "a" x lots + "!"
+    _ = try Regex(#"^((a)+\w)+$"#).firstMatch(in: tainted) // $ redos-vulnerable=
+
+    // BAD
+    // attack string: "bbbbbbbbbb." x lots + "!"
+    _ = try Regex("^(b+.)+$").firstMatch(in: tainted) // $ redos-vulnerable=
+
+    // BAD - all 4 bad combinations of nested * and +
+    // attack string: "a" x lots + "!"
+    _ = try Regex("(a*)*b").firstMatch(in: tainted) // $ redos-vulnerable=
+    _ = try Regex("(a+)*b").firstMatch(in: tainted) // $ redos-vulnerable=
+    _ = try Regex("(a*)+b").firstMatch(in: tainted) // $ redos-vulnerable=
+    _ = try Regex("(a+)+b").firstMatch(in: tainted) // $ redos-vulnerable=
+
+    // GOOD
+    _ = try Regex("(a|b)+").firstMatch(in: tainted)
+
+    // GOOD
+    _ = try Regex(#"(?:[\s;,"'<>(){}|\[\]@=+*]|:(?![/\\]))+"#).firstMatch(in: tainted)
+
+    // BAD?
+    // (no confirmed attack string)
+    _ = try Regex(#"^((?:a{|-)|\w\{)+X$"#).firstMatch(in: tainted) // $ redos-vulnerable=
+    _ = try Regex(#"^((?:a{0|-)|\w\{\d)+X$"#).firstMatch(in: tainted) // $ redos-vulnerable=
+    _ = try Regex(#"^((?:a{0,|-)|\w\{\d,)+X$"#).firstMatch(in: tainted) // $ redos-vulnerable=
+    _ = try Regex(#"^((?:a{0,2|-)|\w\{\d,\d)+X$"#).firstMatch(in: tainted) // $ redos-vulnerable=
+
+    // GOOD
+    _ = try Regex(#"^((?:a{0,2}|-)|\w\{\d,\d\})+X$"#).firstMatch(in: tainted)
+
+    // BAD
+    // attack string: "X" + "a" x lots
+    _ = try Regex(#"X(\u0061|a)*Y"#).firstMatch(in: tainted) // $ redos-vulnerable=
+
+    // GOOD
+    _ = try Regex(#"X(\u0061|b)+Y"#).firstMatch(in: tainted)
+
+    // BAD TODO: we should get this one
+    // attack string: "X" + "a" x lots
+    _ = try Regex(#"X(\x61|a)*Y"#).firstMatch(in: tainted) // $ MISSING: redos-vulnerable=
```

Comment:
unicode escape that you're likely not parsing correctly.  
If you check the [ASCII-table](https://www.asciitable.com/) you'll see that entry 61 (in hexadecimals) is `"a"`.

Reply:
Support for `\x` was in `ParseRegex.qll` but not in `RegexTreeView.qll`.  Now fixed.